twill-core/src/main/java/org/apache/twill/internal/kafka/client/SimpleKafkaConsumer.java

```diff
@@ -273,17 +274,19 @@ private MessageCallback wrapCallback(final MessageCallback callback,
       final AtomicBoolean stopped = new AtomicBoolean();
       return new MessageCallback() {
         @Override
-        public void onReceived(final Iterator<FetchedMessage> messages) {
+        public long onReceived(final Iterator<FetchedMessage> messages, final long startOffset) {
           if (stopped.get()) {
-            return;
+            return startOffset;
           }
-          Futures.getUnchecked(executor.submit(new Runnable() {
+          return Futures.getUnchecked(executor.submit(new Callable<Long>() {
+            long nextOffset = startOffset;
```

Comment:
You don't need this local variable. If the consumer is stopped, it always returns the `startOffset` inside the callable. Otherwise, the callable returns whatever returned by the `callback` delegate.